src/components/spreadsheet/data-fetching/use-fetch-equipment.ts

```diff
@@ -0,0 +1,100 @@
+    );
+
+    const isBuilt = useCallback(
+        (nodeId: string) =>
+            treeNodes?.find(
+                (node) =>
+                    node.id === nodeId && (node.type === NodeType.ROOT || isStatusBuilt(node.data?.globalBuildStatus))
+            ) !== undefined,
+        [treeNodes]
+    );
+
+    const fetchNodesEquipmentData = useCallback(
+        (nodeIds: Set<string>, onFetchingDone: undefined | (() => void)) => {
+            if (studyUuid && currentRootNetworkUuid && currentNodeUuid) {
+                let fetcherPromises: Promise<unknown>[] = [];
+                let spreadsheetEquipmentsByNodes: SpreadsheetEquipmentsByNodes = {
+                    nodesId: [],
+                    equipmentsByNodeId: {},
+                };
+
+                nodeIds.forEach((nodeId) => {
+                    if (currentNodeUuid === nodeId || isBuilt(nodeId)) {
+                        const promise = getFetcher(type)(studyUuid, nodeId as UUID, currentRootNetworkUuid, []);
+                        fetcherPromises.push(promise);
+                        promise
+                            .then((results) => {
+                                let fetchedEquipments = results.flat();
+                                spreadsheetEquipmentsByNodes.nodesId.push(nodeId);
+                                fetchedEquipments = formatEquipments(fetchedEquipments);
+                                spreadsheetEquipmentsByNodes.equipmentsByNodeId[nodeId] = fetchedEquipments;
+                            })
+                            .catch((err) => {
+                                console.error(
+                                    `Fetching error for type ${type.toString()} on node ${nodeId} (${err.message})`
+                                );
+                            });
+                    } else {
+                        // mark un-built node as processed with empty entry/result
+                        spreadsheetEquipmentsByNodes.nodesId.push(nodeId);
+                        spreadsheetEquipmentsByNodes.equipmentsByNodeId[nodeId] = [];
+                    }
+                });
+
+                Promise.all(fetcherPromises)
+                    .then(() => {
+                        dispatch(loadEquipments(type, spreadsheetEquipmentsByNodes));
+                        console.debug(
+                            `Equipment data fetching and dispatch done for ${fetcherPromises.length} built nodes among ${nodeIds.size}`
+                        );
+                    })
+                    .catch((err) => {
+                        console.debug('Equipment data fetching and dispatch NOT done');
+                        snackError({
+                            messageTxt: err.message,
+                            headerId: 'SpreadsheetFetchError',
+                        });
+                    })
+                    .finally(() => {
+                        onFetchingDone && onFetchingDone();
+                    });
```

Comment:
Maybe it's not useful anymore ? We can just call `setIsFetching(false);` after the fetch in use-spreadsheet-equipments

Reply:
when use-spreadsheet-equipments calls fetchNodesEquipmentData(), this is async: we return asap while the promises are going on. The Promises.all.finally may occurs a long time after we return.

If we want to manage correctly the fetching/loading status, we need this callback onFetchingDone.

Reply:
You're right ! Good ! Then can you just replace `(nodeIds: Set<string>, onFetchingDone: undefined | (() => void))` by `(nodeIds: Set<string>, onFetchingDone?: () => void)` so we don't have to explicitely set undefined if we don't want to add a method

Reply:
yes better.